File: script/phys/ammo/Transform.ts
import { Quat } from "./Quat.js";
import { Vec3 } from "./Vec3.js";
import { impl as ammo } from "./context.js";

export class Transform {
    private _position = new Vec3(null);
    get position(): Vec3 {
        this._position.impl = this.impl.getOrigin();
        return this._position;
    }
    set position(value: Vec3) {
        this.impl.setOrigin(value.impl);
    }

    private _rotation = new Quat(null);
    get rotation(): Quat {
        this._rotation.impl = this.impl.getRotation();
        return this._rotation;
    }
    set rotation(value: Quat) {
        this.impl.setRotation(value.impl);
    }

    constructor(public impl = new ammo.btTransform()) { }

    identity() {
        this.impl.setIdentity();
    }
}

File: puttyknife/phys/bullet-end/Transform.cpp
#include <phys/Transform.hpp>
#include <LinearMath/btTransform.h>

phys::Transform *physTransform_new()
{
    return new btTransform();
}

void physTransform_identity(phys::Transform *t)
{
    t->setIdentity();
}

const phys::Vector3 *physTransform_getPosition(phys::Transform *t)
{
    return &t->getOrigin();
}

void physTransform_setPosition(phys::Transform *t, const phys::Vector3 *v3)
{
    t->setOrigin(*v3);
}

const phys::Quat *physTransform_getRotation(phys::Transform *t)
{
    static btQuaternion temp;
    temp = t->getRotation();
    return &temp;
}

void physTransform_setRotation(phys::Transform *t, const phys::Quat *q)
{
    t->setRotation(*q);
}